Write a tutorial for deeplearn.js

chihuahua opened the pull request:
This tutorial takes the reader through making a model to predict the complements of colors within RGB space. We make a simple feed forward network with 3 fully-connected layers.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pair-code/deeplearnjs/15)
<!-- Reviewable:end -->

dsmilkov commented on docs/tutorials/complementary_color_prediction.md:
```diff
@@ -0,0 +1,194 @@
+# Predicting Complementary Colors with deeplearn.js
```
Keep this file in demos/complementary-color-prediction/ - doesn't have to be in docs/tutorials. The website will pick it up and show it on the right side in https://pair-code.github.io/deeplearnjs/docs/tutorials/index.html automatically

chihuahua replied:
Done.

dsmilkov commented on demos/complementary-color-prediction/complementary-color-prediction.html:
```diff
@@ -0,0 +1,136 @@
+<!-- Copyright 2017 Google Inc. All Rights Reserved.
```
Can you run the demo using watch-demo and share the url with us? Thanks!

chihuahua replied:
Done.

dsmilkov commented on demos/complementary-color-prediction/complementary-color-prediction.ts:
```diff
@@ -0,0 +1,355 @@
+/* Copyright 2017 Google Inc. All Rights Reserved.
```
Reminder to run `npm run lint` to check for lint errors. Thank for your patience before we automate this with travis CI.

chihuahua replied:
I get this output

> agent007@agent007:~/deeplearnjs$ npm run lint
> 
> > deeplearn@0.1.0 lint /usr/home/agent007/deeplearnjs
> > tslint -p . --type-check -t verbose
> 
> ERROR: (no-any) /usr/home/agent007/deeplearnjs/demos/homepage/index.ts[30, 63]: Type declaration of 'any' loses type-safety. Consider replacing it with a more precise type, the empty type ('{}'), or suppress this occurrence.
> ERROR: (triple-equals) /usr/home/agent007/deeplearnjs/demos/homepage/index.ts[21, 28]: != should be !==
> 
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! deeplearn@0.1.0 lint: `tslint -p . --type-check -t verbose`
> npm ERR! Exit status 2
> npm ERR! 
> npm ERR! Failed at the deeplearn@0.1.0 lint script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /usr/home/agent007/.npm/_logs/2017-08-09T08_43_14_372Z-debug.log

dsmilkov commented:
Reviewed 2 of 6 files at r2, 1 of 2 files at r3.
Review status: 3 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[npm-debug.log, line 1 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHRJIJQJFgtaK0zv8E:-KrHRJIJQJFgtaK0zv8F:b-6nr0jd) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/npm-debug.log#L1)):*
> ```
> 0 info it worked if it ends with ok
> ```

revert this file

---

*[demos/.DS_Store, line 0 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHRPOJfEfofApBr27Q:-KrHRPOKcEXEUfg2uPnT:b-dojo74) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/.DS_Store#L0)):*
revert this file also

---

*[demos/complementary-color-prediction/complementary-color-prediction.html, line 11 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHReu2G04k54-lEs0B:-KrHReu3lx-MObWZTb_x:bhjie7m) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary-color-prediction.html#L11)):*
> ```HTML
> See the License for the specific language governing permissions and
> limitations under the License.
> ==============================================================================-->
> ```

reduce last line of license to 80 width

---

*[demos/complementary-color-prediction/complementary-color-prediction.ts, line 1 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrDZOh4D-Ahi_pjKncg-r1-1:-KrHrQnrWvpjJLLtksmb:b-fgvqmy) ([raw file](https://github.com/pair-code/deeplearnjs/blob/46acc79034817247811109a787db7b96cbcd61e2/demos/complementary-color-prediction/complementary-color-prediction.ts#L1)):*
<details><summary><i>Previously, chihuahua (Chi Zeng) wrote…</i></summary><blockquote>

I get this output

> agent007@agent007:~/deeplearnjs$ npm run lint
> 
> > deeplearn@0.1.0 lint /usr/home/agent007/deeplearnjs
> > tslint -p . --type-check -t verbose
> 
> ERROR: (no-any) /usr/home/agent007/deeplearnjs/demos/homepage/index.ts[30, 63]: Type declaration of 'any' loses type-safety. Consider replacing it with a more precise type, the empty type ('{}'), or suppress this occurrence.
> ERROR: (triple-equals) /usr/home/agent007/deeplearnjs/demos/homepage/index.ts[21, 28]: != should be !==
> 
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! deeplearn@0.1.0 lint: `tslint -p . --type-check -t verbose`
> npm ERR! Exit status 2
> npm ERR! 
> npm ERR! Failed at the deeplearn@0.1.0 lint script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /usr/home/agent007/.npm/_logs/2017-08-09T08_43_14_372Z-debug.log
</blockquote></details>

Did you run `npm run prep`? Also can you check that you have the right version of the tsc compiler (2.3.4) by running `node_modules/.bin/tsc --version`?

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 3 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHpD44IYtD_a7MC6wi:-KrHpD44IYtD_a7MC6wj:b-lx2kn9) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L3)):*
> may not be perfectly optimized

maybe elaborate a bit on this. Are you thinking about hyper-params or not optimized for speed (not using the library in the best way)?

If hyper-params, you can say smth like
'We didn't spend time tuning the hyper-parameters so there is still lots of room -  we'd love a pull request.'

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 271 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHrjvGQ1f_gFg6U_49:-KrHrjvGQ1f_gFg6U_4A:bliveig) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L271)):*
> ```Markdown
> Based on a few sample colors, it seems like our model of `64 + 32 + 16 = 112` middle-layer nodes fairs alright.
> 
> TODO(chihuahua): Update this image to complementary-color-prediction.png once it is checked in.
> ```

Remove TODO and update the link now? Since it will get checked in this PR.

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 274 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHr07fhRWqTNV4ixmH:-KrHr07fhRWqTNV4ixmI:b-o61qux) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L274)):*
>  offer

s/offer/offers/ (might be wrong, feel free to ignore)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15)*
<!-- Sent from Reviewable.io -->

chihuahua commented:
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[demos/complementary-color-prediction/complementary-color-prediction.html, line 11 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHReu2G04k54-lEs0B:-KrJ5b1_3dcpGqAanGbd:b-cjtuzb) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary-color-prediction.html#L11)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

reduce last line of license to 80 width
</blockquote></details>

Done. Moved closing comment tag to the next line.

---

*[demos/complementary-color-prediction/complementary-color-prediction.ts, line 1 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrDZOh4D-Ahi_pjKncg-r1-1:-KrIu9peytG7hnRIODWX:b-pxpvez) ([raw file](https://github.com/pair-code/deeplearnjs/blob/46acc79034817247811109a787db7b96cbcd61e2/demos/complementary-color-prediction/complementary-color-prediction.ts#L1)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Did you run `npm run prep`? Also can you check that you have the right version of the tsc compiler (2.3.4) by running `node_modules/.bin/tsc --version`?
</blockquote></details>

Indeed, version 2.3.4 ...

agent007@agent007:~/Desktop/deeplearnjs$ node_modules/.bin/tsc --version
Version 2.3.4

I did run `npm run prep`. I tried again and think I get the same output ... odd. Here is the output of the log file. I might have some time later to dig deeper.

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/home/agent007/.nvm/versions/node/v8.3.0/bin/node',
1 verbose cli   '/usr/local/home/agent007/.nvm/versions/node/v8.3.0/bin/npm',
1 verbose cli   'run',
1 verbose cli   'lint' ]
2 info using npm@5.3.0
3 info using node@v8.3.0
4 verbose run-script [ 'prelint', 'lint', 'postlint' ]
5 info lifecycle deeplearn@0.1.0~prelint: deeplearn@0.1.0
6 info lifecycle deeplearn@0.1.0~lint: deeplearn@0.1.0
7 verbose lifecycle deeplearn@0.1.0~lint: unsafe-perm in lifecycle true
8 verbose lifecycle deeplearn@0.1.0~lint: PATH: /usr/local/home/agent007/.nvm/versions/node/v8.3.0/lib/node_modules/npm/bin/node-gyp-bin:/usr/local/home/agent007/Desktop/deeplearnjs/node_modules/.bin:/usr/local/home/agent007/.nvm/versions/node/v8.3.0/bin:/usr/local/home/agent007/anaconda2/bin:/usr/lib/google-golang/bin:/usr/local/buildtools/java/jdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/src/head/depot/google3/third_party/gpus/cuda_8_0/bin
9 verbose lifecycle deeplearn@0.1.0~lint: CWD: /usr/local/home/agent007/Desktop/deeplearnjs
10 silly lifecycle deeplearn@0.1.0~lint: Args: [ '-c', 'tslint -p . --type-check -t verbose' ]
11 silly lifecycle deeplearn@0.1.0~lint: Returned: code: 2  signal: null
12 info lifecycle deeplearn@0.1.0~lint: Failed to exec lint script
13 verbose stack Error: deeplearn@0.1.0 lint: `tslint -p . --type-check -t verbose`
13 verbose stack Exit status 2
13 verbose stack     at EventEmitter.<anonymous> (/usr/local/home/agent007/.nvm/versions/node/v8.3.0/lib/node_modules/npm/lib/utils/lifecycle.js:289:16)
13 verbose stack     at emitTwo (events.js:125:13)
13 verbose stack     at EventEmitter.emit (events.js:213:7)
13 verbose stack     at ChildProcess.<anonymous> (/usr/local/home/agent007/.nvm/versions/node/v8.3.0/lib/node_modules/npm/lib/utils/spawn.js:40:14)
13 verbose stack     at emitTwo (events.js:125:13)
13 verbose stack     at ChildProcess.emit (events.js:213:7)
13 verbose stack     at maybeClose (internal/child_process.js:927:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
14 verbose pkgid deeplearn@0.1.0
15 verbose cwd /usr/local/home/agent007/Desktop/deeplearnjs
16 verbose Linux 4.4.0-83-generic
17 verbose argv "/usr/local/home/agent007/.nvm/versions/node/v8.3.0/bin/node" "/usr/local/home/agent007/.nvm/versions/node/v8.3.0/bin/npm" "run" "lint"
18 verbose node v8.3.0
19 verbose npm  v5.3.0
20 error code ELIFECYCLE
21 error errno 2
22 error deeplearn@0.1.0 lint: `tslint -p . --type-check -t verbose`
22 error Exit status 2
23 error Failed at the deeplearn@0.1.0 lint script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 2, true ]
```

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 3 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHpD44IYtD_a7MC6wi:-KrJ29ieQejgdZBCnEGR:bjto8ld) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L3)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

> may not be perfectly optimized

maybe elaborate a bit on this. Are you thinking about hyper-params or not optimized for speed (not using the library in the best way)?

If hyper-params, you can say smth like
'We didn't spend time tuning the hyper-parameters so there is still lots of room -  we'd love a pull request.'
</blockquote></details>

Done. Right - I meant that this tutorial did not optimize for hyperparameters (number of layers, sizes of layers). If you have any feedback on using the library better - let me know! 

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 271 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHrjvGQ1f_gFg6U_49:-KrJ2miHU43hHchBhBSc:bf9e50w) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L271)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

Remove TODO and update the link now? Since it will get checked in this PR.
</blockquote></details>

I removed the TODO and verified that the markdown should look correct:

![RouVWhA1GDN.png](https://files.reviewableusercontent.io/images/reviewable/4221553/H3f-e0Hj0wnHIJjV0s5bJrbI/RouVWhA1GDN.png)

The markdown preview currently does not show the image because the URL is relative to the PAIR-code organization, and the image has not yet been checked into the PAIR-code/deeplearnjs repo yet.

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 274 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHr07fhRWqTNV4ixmH:-KrJ2ZQG05A74tSGqCHj:bx5dcfa) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L274)):*
> s within it offer a 
Done.

Also, to avoid bloating this tutorial, I sort of glossed over the logic for managing the UI. That also makes the tutorial focus on deeplearn.js. Let me know if you have thoughts on that.

---

*[npm-debug.log, line 1 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHRJIJQJFgtaK0zv8E:-KrJ4olcjVuk7WlBLiEF:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/npm-debug.log#L1)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

revert this file
</blockquote></details>

Done.

---

*[demos/.DS_Store, line  at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHRPOJfEfofApBr27Q:-KrJ4poGPZKbxspZzi1Y:b-896fix) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/.DS_Store#L)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

revert this file also
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15)*
<!-- Sent from Reviewable.io -->

dsmilkov commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 6 files at r2, 1 of 1 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 271 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHrjvGQ1f_gFg6U_49:-KrJB4a1gWn88fezHG7d:bbu3max) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L271)):*
<details><summary><i>Previously, chihuahua (Chi Zeng) wrote…</i></summary><blockquote>

I removed the TODO and verified that the markdown should look correct:

![RouVWhA1GDN.png](https://files.reviewableusercontent.io/images/reviewable/4221553/H3f-e0Hj0wnHIJjV0s5bJrbI/RouVWhA1GDN.png)

The markdown preview currently does not show the image because the URL is relative to the PAIR-code organization, and the image has not yet been checked into the PAIR-code/deeplearnjs repo yet.
</blockquote></details>

I think you can do relative paths in markdown links, .e.g 
[Predicting Complementary Colors](complementary-color-prediction.png)
also when linking the other articles from this article.

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 274 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHr07fhRWqTNV4ixmH:-KrJBXx6_WEW2IJgDMA-:b-aekegx) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L274)):*
<details><summary><i>Previously, chihuahua (Chi Zeng) wrote…</i></summary><blockquote>

> s within it offer a 
Done.

Also, to avoid bloating this tutorial, I sort of glossed over the logic for managing the UI. That also makes the tutorial focus on deeplearn.js. Let me know if you have thoughts on that.
</blockquote></details>

I think it's good. Short and sweet. We can get feedback from our users and improve later.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-:-KrJCb-1N-4oQ6OpY7CD:bnfp4nl)*
<!-- Sent from Reviewable.io -->

chihuahua commented:
Review status: 3 of 5 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[demos/complementary-color-prediction/complementary_color_prediction.md, line 271 at r4](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15#-KrHrjvGQ1f_gFg6U_49:-KrNtm8-uUjCS1MuemdT:b-7b41) ([raw file](https://github.com/pair-code/deeplearnjs/blob/fffa88ba8fdb4dc44e95b27d6bb9aa8eb7ce86c6/demos/complementary-color-prediction/complementary_color_prediction.md#L271)):*
<details><summary><i>Previously, dsmilkov (Daniel Smilkov) wrote…</i></summary><blockquote>

I think you can do relative paths in markdown links, .e.g 
[Predicting Complementary Colors](complementary-color-prediction.png)
also when linking the other articles from this article.
</blockquote></details>

Indeed, relative links work! Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/15)*
<!-- Sent from Reviewable.io -->

chihuahua commented:
I modified relative links. Feel free to merge at any time.

One problem I realized is that, sometimes, certain weight values are initialized to 0, and then those color channels just stay at 0. Like this:

![usavox2kr9j](https://user-images.githubusercontent.com/4221553/29245073-f828e6fe-7f81-11e7-9a54-8f026c65c0f4.png)

I so far lack a solution for this. Refreshing the page usually allows for normal training, but it's unfortunate that initialization of weights matters.

chihuahua commented:
Interestingly, despite having a channel stuck at 0, the predictions are often not *that* bad:

![blue-channel-at-0](https://user-images.githubusercontent.com/4221553/29245125-4acd7c92-7f84-11e7-8b3e-b815ebc9bdba.png)

dsmilkov commented:
Ha, that's very interesting about the weights going to 0.

Wohoo, thanks for writing this awesome tutorial!
